### PR TITLE
New expression editor v1

### PR DESCRIPTION
**Pull request by RyanStonebraker**
- Creates new fullscreen expression editor
  - Fullscreen editor can be accessed by clicking solo mode toggle in popup text editor
- Modularizes expression text area into separate component so it can be used in popup text editor and full screen
- Introduces debug mode where expressions can be run in full or just a highlighted sub section without saving
  - If a highlighted sub section is run, the background of the run text is changed to black. As soon as any changes are made, this goes away and a warning icon appears in the top right letting the user know the visualization is out of sync
- Expressions (or subset expressions) will automatically show in visualization widget when run
  - Widgets that are available in the preview mode are:
    - Context Image
    - Binary Plot
    - Ternary Plot
    - Chord Diagram
    - Histogram
- Introduces new PMC data grid widget
  - Shows average, min, and max for all PMC expression values
  - Hovering is synced across widgets
  - Selection is shown with a blue background for the cell and clicking on a cell will toggle the PMC's selection
  - Widget can be toggled to take up the entire right side
  - PMC and Expression Values can be exported from export dialog
- Adds hotkeys
  - Cmd/Ctrl + "S": Save
  - Cmd/Ctrl + Enter: Run Expression
  - Cmd/Ctrl + Alt/Option + Enter: Run Highlighted Expression or if no selection, run expression to line
  - Cmd/Ctrl + "/": Toggle line comment
-  Makes autocomplete/help menu toggleable

## TODO
- [x] Need error overlay for PMC data grid when no valid data is available
  - This design was originally meant for variable level debugging - we went in a different direction and don't have this so this might change?
- [x] Need to merge in new Lua functionality

### CODE REVIEW

File: client/src/app/UI/context-image-view-widget/context-image-view-widget.component.html [hunk 1]
```diff
@@ -31,7 +31,7 @@
 
 <div fxLayout="column" class="panel outer-panel" fxFill> <!-- fxFill here fills the parent -->
     <div fxLayout="row" fxLayoutAlign="space-between center" class="panel-title">
-        <widget-switcher *ngIf="allowSwitchingTools" [activeSelector]='thisSelector' [widgetPosition]='widgetPosition'></widget-switcher>
+        <widget-switcher *ngIf="allowSwitchingTools" [activeSelector]='thisSelector' [widgetPosition]='widgetPosition' [previewMode]="this.previewExpressionIDs.length > 0"></widget-switcher>
```
Comment (pnemere) [inline client/src/app/UI/context-image-view-widget/context-image-view-widget.component.html:34]: maybe use isPreviewMode here?

File: client/src/app/routes/dataset/analysis/analysis.component.ts [hunk 2]
```diff
@@ -327,10 +327,14 @@ export class AnalysisComponent implements OnInit, OnDestroy
     private createTopRowComponents(selectors: string[])
     {
         // Should be 2 components... if not, show the default 2
-        if(selectors.length != 2)
+        if(selectors.length < 2)
         {
             selectors = [ViewStateService.widgetSelectorContextImage, ViewStateService.widgetSelectorSpectrum];
         }
+        else if(selectors.length > 2)
```
Comment (pnemere) [inline client/src/app/routes/dataset/analysis/analysis.component.ts:334]: What happens if selectors > 2? Are we going to allow more widgets up top or is this a bug?

File: client/src/app/routes/dataset/code-editor/code-editor.component.html [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                [columnCount]="15"
+                (onToggleSolo)="onTogglePMCDataGridSolo($event)"
+            ></pmc-data-grid>
+        </div>
+    </div>
+    <div class="editor-footer">
+        <div class="text-container">
+            <mat-select [(ngModel)]="isLua">
+                <mat-option [value]="true">Lua</mat-option>
+                <mat-option [value]="false">Pixlang</mat-option>
```
Comment (pnemere) [inline client/src/app/routes/dataset/code-editor/code-editor.component.html:126]: 😆 Pixlang, why not!

File: client/src/app/services/view-state.service.ts [hunk 4]
```diff
@@ -1065,6 +1065,11 @@ export class ViewStateService
 
                         this._viewSelectorsAreDefaults = true;
                     }
+                    else if(state.analysisLayout.bottomWidgetSelectors.length > 4)
+                    {
+                        // If we have more than 4 widgets along the bottom, we'll just use the first 4
+                        state.analysisLayout.bottomWidgetSelectors = state.analysisLayout.bottomWidgetSelectors.slice(0, 4);
```
Comment (pnemere) [inline client/src/app/services/view-state.service.ts:1071]: Same question, what happens to the ones > 4?

File: client/src/app/services/view-state.service.ts [hunk 5]
```diff
@@ -1296,7 +1301,7 @@ export class ViewStateService
                 ), "underspectrum1");
 
                 // Set top widgets to defaults
-                this._viewState.analysisLayout.topWidgetSelectors = [ViewStateService.widgetSelectorContextImage, ViewStateService.widgetSelectorSpectrum];
+                this._viewState.analysisLayout.topWidgetSelectors = [ViewStateService.widgetSelectorContextImage, ViewStateService.widgetSelectorSpectrum, ViewStateService.widgetSelectorTernaryPlot];
```
Comment (pnemere) [inline client/src/app/services/view-state.service.ts:1304]: Oh, we're putting a ternary up top but hiding it? Is this a hack to remember the state?

File: client/src/app/UI/expression-editor/expression-text-editor/expression-text-editor.component.ts [hunk 6]
```diff
@@ -0,0 +1,954 @@
...
+            cm.setSelection(cursor, cursor);
+        };
+        
+        if(this.isWindows)
+        {
+            cmObj.setOption("extraKeys", {
+                "Ctrl-/": toggleComment,
+                "Ctrl-Enter": (cm) =>
+                {
+                    resetMarks(cm);
```
Comment (pnemere) [inline client/src/app/UI/expression-editor/expression-text-editor/expression-text-editor.component.ts:190]: Coming from Visual Studio back in the day, I'd want F7 to run and F5 to debug too :) Just kidding though, don't change it!

**Review by pnemere — APPROVED** [review]
All looks great, added a few comments, some are easy fixes! I'll mark it as approved though

File: client/src/app/routes/dataset/analysis/analysis.component.ts (hunk 2 above)
Comment (RyanStonebraker) [reply client/src/app/routes/dataset/analysis/analysis.component.ts:334]: When I was adding the new preview widget, I got it into a state for a bit where there were 3 selectors in the top row and this got saved to my view state, which caused it previously to just revert to the default selectors (previous logic was to do that if `length != 2`). I changed it to instead only revert to the default selectors if there were less than 2 selectors stored in the user view state, or if there were more than 3, then just use the first 2. While this probably won't affect anyone, I figured if someone had a malformed view state like I did, it'd be better to recover the valid parts than do a complete reset 🤷‍♂️

File: client/src/app/routes/dataset/code-editor/code-editor.component.html (hunk 3 above)
Comment (RyanStonebraker) [reply client/src/app/routes/dataset/code-editor/code-editor.component.html:126]: Happy to change it, but I was using `Pixlise Expression Language` before and that truncated the text weird after it was selected so needed something shorter 😅

File: client/src/app/services/view-state.service.ts (hunk 5 above)
Comment (RyanStonebraker) [reply client/src/app/services/view-state.service.ts:1305]: Yeah the 3rd top widget selector is the preview container - we could just default this to something if we don't want to store it, but the widget selectors are just stored as a list of strings, so this seemed like the easiest way to get it in.

**Pull request merged**